sdk/core/Azure.Core.Experimental/src/Serialization/JsonObjectSerializer.cs

```diff
@@ -59,5 +67,85 @@ public override async ValueTask<object> DeserializeAsync(Stream stream, Type ret
+
+            if (!_cache.TryGetValue(memberInfo.ReflectedType, out MemberInfo[] members))
+            {
+                members = GetMembers(memberInfo.ReflectedType).ToArray();
+            }
+
+            foreach (MemberInfo member in members)
+            {
+                if (member == memberInfo)
+                {
+                    return GetPropertyName(memberInfo);
+                }
+            }
+
+            return null;
+        }
+
+        private static IEnumerable<MemberInfo> GetMembers(Type type)
+        {
+            // Mimics property enumeration based on:
+            // * https://github.com/dotnet/corefx/blob/v3.1.0/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L130-L191
+            // * TODO: https://github.com/dotnet/runtime/blob/dc8b6f90/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L147-L155
+
+            PropertyInfo[] properties = type.GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+            foreach (PropertyInfo propertyInfo in properties)
+            {
+                // Ignore indexers.
+                if (propertyInfo.GetIndexParameters().Length > 0)
+                {
+                    continue;
+                }
+
+                // Only support public getters and/or setters.
+                if (propertyInfo.GetMethod?.IsPublic == true ||
+                    propertyInfo.SetMethod?.IsPublic == true)
+                {
+                    if (propertyInfo.GetCustomAttribute<JsonIgnoreAttribute>() != null)
+                    {
+                        continue;
+                    }
+
+                    // Ignore - but do not assert correctness - for JsonExtensionDataAttribute based on
+                    // https://github.com/dotnet/corefx/blob/v3.1.0/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L244-L261
+                    if (propertyInfo.GetCustomAttribute<JsonExtensionDataAttribute>() != null)
+                    {
+                        continue;
+                    }
+
+                    // No need to validate collisions since they are based on the serialized name.
+                    yield return propertyInfo;
+                }
+            }
+        }
+
+        private string GetPropertyName(MemberInfo memberInfo)
+        {
+            // Mimics property name determination based on
+            // https://github.com/dotnet/runtime/blob/dc8b6f90/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs#L53-L90
+
+            JsonPropertyNameAttribute nameAttribute = memberInfo.GetCustomAttribute<JsonPropertyNameAttribute>(false);
+            if (nameAttribute != null)
+            {
+                return nameAttribute.Name
+                    ?? throw new InvalidOperationException($"The JSON property name for '{memberInfo.DeclaringType}.{memberInfo.Name}' cannot be null.");
```

Comment:
I think it'd be better to return null than validate behavior from System.Text.Json in these cases.

Reply:
That would mean the serialized results wouldn't match the outcome of this implementation, which will lead to problems in some cases. As with the `ArgumentNullException` above, this indicates a developer error that should be resolved.